### PR TITLE
[WFLY-6545] Add support for a 'disallow-doctype-decl' attribute to the JSF subsystem

### PR DESCRIPTION
**Pull request by fjuma**
https://issues.jboss.org/browse/WFLY-6545

### CODE REVIEW

File: jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFSubsystemAdd.java [hunk 1]
```diff
@@ -50,16 +50,19 @@
...
     protected void populateModel(ModelNode operation, ModelNode model) throws OperationFailedException {
         JSFResourceDefinition.DEFAULT_JSF_IMPL_SLOT.validateAndSet(operation, model);
+        JSFResourceDefinition.DISALLOW_DOCTYPE_DECL.validateAndSet(operation, model);
     }
 
     @Override
-    protected void performBoottime(OperationContext context, ModelNode operation, final ModelNode model) {
+    protected void performBoottime(OperationContext context, ModelNode operation, final ModelNode model) throws OperationFailedException {
+        final Boolean disallowDoctypeDecl = model.hasDefined(JSFResourceDefinition.DISALLOW_DOCTYPE_DECL_ATTR_NAME) ?
+                JSFResourceDefinition.DISALLOW_DOCTYPE_DECL.resolveModelAttribute(context, model).asBoolean() : null;
```
Comment (bstansberry) [inline jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFSubsystemAdd.java]:
```
final Boolean disallowDoctypeDecl = JSFResourceDefinition.DISALLOW_DOCTYPE_DECL.resolveModelAttribute(context, model).asBooleanOrNull();
```

File: jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFExtension.java [hunk 2]
```diff
@@ -69,13 +70,19 @@ public void initialize(final ExtensionContext context) {
         JSFLogger.ROOT_LOGGER.debug("Activating JSF(Mojarra) Extension");
         final SubsystemRegistration subsystem = context.registerSubsystem(SUBSYSTEM_NAME, CURRENT_MODEL_VERSION);
         subsystem.registerSubsystemModel(JSFResourceDefinition.INSTANCE);
-        subsystem.registerXMLElementWriter(JSFSubsystemParser_1_0.INSTANCE);
+        subsystem.registerXMLElementWriter(JSFSubsystemParser_1_1.INSTANCE);
+
+        // Transformers
+        if (context.isRegisterTransformers()) {
+            JSFResourceDefinition.registerTransformers(subsystem);
```
Comment (bstansberry) [inline jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFExtension.java]: There's a new ServiceLoader based way of registering transformers that should be used -- see implementations of org.jboss.as.controller.transform.ExtensionTransformerRegistration

File: jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFMetadataProcessor.java [hunk 3]
```diff
@@ -34,10 +46,28 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
...
     public void undeploy(DeploymentUnit context) {
     }
+
+    private void setContextParameterIfAbsent(final JBossWebMetaData webMetaData, final String name, final String value) {
+        List<ParamValueMetaData> contextParams = webMetaData.getContextParams();
+        if (contextParams == null) {
+            contextParams = new ArrayList<ParamValueMetaData>();
+            webMetaData.setContextParams(contextParams);
+        }
+        boolean paramFound = contextParams.stream().anyMatch(param -> name.equals(param.getParamName()) && param.getParamValue() != null);
```
Comment (bstansberry) [inline jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFMetadataProcessor.java]: Please switch to a basic for loop. Lambdas are expensive in terms of class storage space.

**Review by bstansberry — CHANGES_REQUESTED** [review]

**Comment by fjuma** [comment]
Fixed and rebased. Note that the upgrade to jsf-impl 2.2.13.SP2 is no longer needed since master is already on jsf-impl 2.2.13.SP4.

**Comment by fjuma** [comment]
retest this please

**Comment by fjuma** [comment]
retest this please

**Review by bstansberry — APPROVED** [review]

**Pull request merged**